### ISSUE TITLE
预训练中构建全词mask问题

### ISSUE THREAD

**Issue by 1073521013** [bojone/bert4keras#386]
提问时请尽可能提供如下信息：

### 基本信息
- 你使用的**操作系统**: 
- 你使用的**Python**版本: 
- 你使用的**Tensorflow**版本: 
- 你使用的**Keras**版本: 
- 你使用的**bert4keras**版本: 
- 你使用纯**keras**还是**tf.keras**: 
- 你加载的**预训练模型**:

### 核心代码
```python
# 请在此处贴上你的核心代码。
# 请尽量只保留关键部分，不要无脑贴全部代码。
```
    def sentence_process(self, text):
        """单个文本的处理函数
        流程：分词，然后转id，按照mask_rate构建全词mask的序列
              来指定哪些token是否要被mask
        """
        words = self.word_segment(text)
        rands = np.random.random(len(words))

        token_ids, mask_ids = [], []
        for rand, word in zip(rands, words):
            word_tokens = self.tokenizer.tokenize(text=word)[1:-1]
            word_token_ids = self.tokenizer.tokens_to_ids(word_tokens)
            token_ids.extend(word_token_ids)

            if rand < self.mask_rate:
                word_mask_ids = [
                    self.token_process(i) + 1 for i in word_token_ids
                ]
### 输出信息
```shell
# 请在此处贴上你的调试输出
```

### 自我尝试
不管什么问题，请先尝试自行解决，“万般努力”之下仍然无法解决再来提问。此处请贴上你的努力过程。

苏神你好，麻烦咨询一下预训练数据处理的两个问题：
1 请问为什么全词mask词汇非首字没有人工加上”##“标记再去映射id呢？
2 请问15%比例替换后为什么有个+1操作呢？
感谢

**Comment by bojone** [comment]
1、为什么要加？
2、+1只是编程技巧，后面会重新减去1
3、为什么Input-Token是token_ids？token_ids是原始没有被[mask]过的序列，你输入token_ids，想预测什么？

请融会贯通上下文来理解，不要只看局部。

**Comment by 1073521013** [comment]
1、transformer原版的实现确实都没加，brightmart在预测的部分加了，个人觉得这块加前缀优势可以学到细粒度信息，但是下游任务一般wordpicse分词所以存在gap，我在3个下游任务做了几个测试，有两个任务加前缀f1略高半个百分点，可能需要更多实验验证吧。
2、方便说下啥技巧吗
3、抱歉，这个问题是我疏忽了。

**Comment by bojone** [comment]
1、你的意思是同时保留“的”和“##的”，然后下游任务也保留词信息？那不是依赖分词工具了？没什么意思吧？

2、技巧是对于编程者来说的，既然你认识不到这个技巧，你可以不接受这个写法，而自行构思符合自己认知的写法。（这不是抬杠，这是事实，因为对于每个人来说“显然成立”的东西都不一样。）

**Comment by 1073521013** [comment]
1 下游任务不会保留“##”前缀信息，也就是不会用到分词，所以和训练存在不一致的情况, 就跟预测“<mask>”标记似的，预训练和下游任务也是存在不一致情况，但是个人感觉预训练过程增加“##”确实可以学到细粒度信息吧，比如“苹果”的”##果“和“果然”的“果”特征学习还是有利于模型收敛的。
2 好的，非常感谢耐心解答
